db/c.cc

```diff
@@ -3156,6 +3234,12 @@ void rocksdb_options_set_merge_operator(
   opt->rep.merge_operator = std::shared_ptr<MergeOperator>(merge_operator);
 }
 
+void rocksdb_options_add_event_listener(
+    rocksdb_options_t* options, rocksdb_event_listener_t* event_listener) {
+  std::shared_ptr<EventListener> el(event_listener);
+  options->rep.listeners.emplace_back(el);
+}
+
```

Comment:
The shared_ptr will take ownership of the event_listener pointer, this opens the door for misuse if anyone calls `rocksdb_event_listener_destroy` on that pointer after this point. 

I see the options as:
- Make `rocksdb_event_listener_t` own the shared_ptr and hand it down here. Therefore, the ownership is truly shared (probably best)
- Add a comment on this function to make the ownership semantics very clear.

Reply:
I see the problem! I'm effectively leaking the `EventListener` when setting it, with no way to get it back. I've solved it by owning the `shared_ptr` in the `rocksdb_event_listener_t` as you suggested - but that required returning that back to Rust-land in a holder, otherwise we have no way of retaining it. 

The caller has to retain this holder, not just the underlying `EventListener` impl, but the upshot is that we can easily implement `Options::remove_event_listener` if required.

Reply:
>The shared_ptr will take ownership of the event_listener pointer, this opens the door for misuse if anyone calls rocksdb_event_listener_destroy on that pointer after this point.

Despite my suggestion. I found that it's a common pattern in this particular ffi interface (rocksdb), so I'd be okay with keeping it aligned with their conventions. Essentially, if you pass the listener to add_listener, it'll take ownership of the memory, if you didn't, you'll have to call listener_destroy yourself.

Reply:
Okie dokie! Then we no longer need the `rocksdb_event_listener_inner_t` type nor the `EventListenerMustOutliveDB` struct on the Rust side.